### ISSUE TITLE
Add 'CitiBank Polecenie Zaplaty' file format

### ISSUE THREAD

**Issue by systopia** [Project60/org.project60.sepa#298]
As described in https://github.com/Project60/org.project60.sepa/pull/297, some folks want to add a new file format. However, it's obviously not done by adding the option value.

The current system has a predefined algorithm, but if you (@michau or @scardinius) can provide a smarty template for that, it wouldn't be too much work to integrate.

Also, the file format would need a sensible name, i.e. the format specification instead of the bank's name.

**Comment by systopia** [comment]
@michau / @scardinius: let us know if you're interested in a cooperation.

**Comment by michau** [comment]
@systopia Yes, we're interested in the cooperation. :-) Right now we're at the stage where we're learning CiviSEPA code and trying to find out what needs to be done in order to use it for CitiBank direct debit mechanism.

First of all, let's scratch the option_value SQL code in the comment - we've put it in the comment for reference, but it's practically irrelevant to broader issue. :-) Sorry for confusion. :-)

In general, seems like CitiBank in Poland has slightly different way of handling direct debit than SEPA standard (format wise), but whole process stays more or less the same. To be more specific: you submit a list of mandates for verification in specific format (format 1), receive mandate verification status info, than submit transaction requests normally in another format (format 2). Both are text with delimiters, with headers and footers.

So far, we identified following things that we would need to do in order to be able to use CiviSEPA:

1. implement ability to plug in CitiBank specific format
2. implement ability to export mandate verification file (naming convention for this file includes Client's codename given by Bank, so we'll need a format specific settings field here)
3. implement ability to parse and import both mandate verification status file and transactions status files from bank
4. potentially allow handling multiple currencies (PLN transations handled here)

We have much longer list of potential smaller changes and improvements, but leaving these aside for now to focus on base issue - adding "pluggable formats".

So - the pull request accompanying #297 (see: https://github.com/scardinius/org.project60.sepa/commit/551c56789b80ef6db88847d8a6e4f6b679adccc6) does not consist of option_value changes only - this is proof of concept refactoring trying to sketch out the concept of pluggable formats. Please take a quick look and let us know if that's something that would. Currently implemented XML format would stay as default and work as before.

We cannot fund this feature, but we'll gladly implement it and contribute the code, together with other improvements, testing effort, maybe even some significant documentation improvements.

**Comment by systopia** [comment]
@michau Sorry, I was fooled by the comment - I really didn't see that the PR was much bigger :-)

Anyway, I like the idea of the pluggable file formats, and your solution seems solid on first glance. I will try to merge it into 1.3, but please understand that if it takes up too much of my time, we would need to find some funding for this.

With respect to your other plans (especially multi currency), I'm looking forward to your PRs. Please be very thorough there, since half-baked solutions would destabilize the system for the other users - and we won't be able pull it :)

**Comment by scardinius** [comment]
@systopia We're glad you're interested in our contribution. 

Hold on, please. We're still working on this functionality and we will prepare proper PR. This was just a proof of concept.

**Comment by scardinius** [comment]
Changes for pluggable file formats with custom filename:

* Templates are moved to new directory: `formats`
* Each format is located in separate directory
* Directory name is sanitized (pain.008.003.02 COR1 -> pain_008_003_02_COR1)
* In directory are placed smarty templates with new filenames
* In directory is placed class inherited from abstract class `CRM_Sepa_Logic_Format` which purchases default values for file prefix and filename
* Citibank format is added to install procedure
* Citibank formats have not yet prepared (only test's content)
* Sanitize method is located in `CRM_Utils_SepaOptionGroupTools` class

What do you think about such solution?

**Comment by systopia** [comment]
Sounds good. I'll have a closer look, once v1.2 is released.

**Comment by systopia** [comment]
Well, since 1.2 is taking more time to be released than expected, I changed the master to 1.3dev and merged your PR. Happy testing :)

**Comment by systopia** [comment]
It doesn't seem to work properly... the CitiBank file created on my system looks like this:
````
transaction-header
transaction-details
transaction-footer
```
The old format however seems to be fine, good job!

**Comment by scardinius** [comment]
@systopia thx for merging pluggable formats.

Don't worry, CitiBank templates include only testing content. I'm going to working on templates in a few days.

**Comment by systopia** [comment]
Ok, just remember to create a PR when you're done.

**Comment by systopia** [comment]
This is now in ``master`` with #314, and should be released with ``1.2.6``
